File: frontend/lib/api/services.ts
import { api } from "./client";

export interface Service {
  id: string;
  name: string;
  slug: string;
  short_description?: string | null;
  description?: string | null;
  category:
    | "WEB_DEVELOPMENT"
    | "SOFTWARE_DEVELOPMENT"
    | "BUSINESS_AUTOMATION"
    | "AI_SYSTEMS"
    | "DATA_ANALYTICS"
    | "UI_UX"
    | "MAINTENANCE"
    | "CONSULTING"
    | "OTHER";
  subcategory?: string | null;
  status: "DRAFT" | "ACTIVE" | "PAUSED" | "ARCHIVED";
  delivery_model:
    | "FIXED_PROJECT"
    | "CUSTOM_QUOTE"
    | "SUBSCRIPTION"
    | "RETAINER"
    | "HOURLY"
    | "CONSULTATION";
  pricing_model: "FIXED" | "STARTING_AT" | "RANGE" | "CUSTOM" | "NOT_SET";
  base_price?: number | null;
  price_min?: number | null;
  price_max?: number | null;
  currency: string;
  estimated_duration_days?: number | null;
  is_featured: boolean;
  is_active: boolean;
  features: string[];
  requirements: string[];
  target_business_types: string[];
  archived_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ServiceCreateInput {
  name: string;
  slug?: string;
  short_description?: string;
  description?: string;
  category: Service["category"];
  subcategory?: string;
  status?: Service["status"];
  delivery_model?: Service["delivery_model"];
  pricing_model?: Service["pricing_model"];
  base_price?: number;
  price_min?: number;
  price_max?: number;
  currency?: string;
  estimated_duration_days?: number;
  is_featured?: boolean;
  is_active?: boolean;
  features?: string[];
  requirements?: string[];
  target_business_types?: string[];
}

export interface ServiceUpdateInput extends Partial<ServiceCreateInput> {}

export interface LeadService {
  id: string;
  lead_id: string;
  service_id: string;
  relationship_type: "CONSIDERED" | "RECOMMENDED" | "SELECTED" | "REJECTED";
  source: "HUMAN" | "RULE" | "AI" | "IMPORT" | "OTHER";
  notes?: string | null;
  created_at: string;
  updated_at: string;
  service?: Service | null;
}

export interface ServiceListParams {
  page?: number;
  page_size?: number;
  search?: string;
  category?: string;
  status?: string;
  delivery_model?: string;
  pricing_model?: string;
  is_featured?: boolean;
  is_active?: boolean;
  sort?: string;
  order?: "asc" | "desc";
}

export interface PaginatedServicesResponse {
  data: Service[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
}

export async function listServices(params: ServiceListParams = {}): Promise<PaginatedServicesResponse> {
  const query = new URLSearchParams();
  if (params.page) query.append("page", params.page.toString());
  if (params.page_size) query.append("page_size", params.page_size.toString());
  if (params.search) query.append("search", params.search);
  if (params.category) query.append("category", params.category);
  if (params.status) query.append("status", params.status);
  if (params.delivery_model) query.append("delivery_model", params.delivery_model);
  if (params.pricing_model) query.append("pricing_model", params.pricing_model);
  if (params.is_featured !== undefined) query.append("is_featured", params.is_featured.toString());
  if (params.is_active !== undefined) query.append("is_active", params.is_active.toString());
  if (params.sort) query.append("sort", params.sort);
  if (params.order) query.append("order", params.order);

  const queryString = query.toString() ? `?${query.toString()}` : "";
  return api<PaginatedServicesResponse>(`/services${queryString}`);
}

export async function getService(id: string): Promise<{ data: Service }> {
  return api<{ data: Service }>(`/services/${id}`);
}

export async function getServiceBySlug(slug: string): Promise<{ data: Service }> {
  return api<{ data: Service }>(`/services/slug/${slug}`);
}

export async function createService(input: ServiceCreateInput): Promise<{ data: Service }> {
  return api<{ data: Service }>("/services", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function updateService(id: string, input: ServiceUpdateInput): Promise<{ data: Service }> {
  return api<{ data: Service }>(`/services/${id}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export async function archiveService(id: string): Promise<{ data: Service }> {
  return api<{ data: Service }>(`/services/${id}/archive`, {
    method: "POST",
  });
}

export async function restoreService(id: string): Promise<{ data: Service }> {
  return api<{ data: Service }>(`/services/${id}/restore`, {
    method: "POST",
  });
}

// Lead-Service APIs
export async function listLeadServices(leadId: string): Promise<{ data: LeadService[] }> {
  return api<{ data: LeadService[] }>(`/leads/${leadId}/services`);
}

export async function addServiceToLead(
  leadId: string,
  input: {
    service_id: string;
    relationship_type?: LeadService["relationship_type"];
    source?: LeadService["source"];
    notes?: string;
  }
): Promise<{ data: LeadService }> {
  return api<{ data: LeadService }>(`/leads/${leadId}/services`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function removeServiceFromLead(
  leadId: string,
  serviceId: string
): Promise<{ data: { success: boolean; message: string } }> {
  return api<{ data: { success: boolean; message: string } }>(`/leads/${leadId}/services/${serviceId}`, {
    method: "DELETE",
  });
}

File: backend/app/api/v1/services.py
import math
import uuid
from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.common import DataResponse, PaginatedResponse, PaginationMeta
from app.schemas.service import (
    LeadServiceCreate,
    LeadServiceResponse,
    ServiceCreate,
    ServiceResponse,
    ServiceUpdate,
)
from app.services.service import ServiceCatalogService

router = APIRouter(tags=["Service Catalog"])


@router.post("/services", response_model=DataResponse[ServiceResponse], status_code=201)
def create_service(
    data: ServiceCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ServiceCatalogService.create_service(db, data)
    return {"data": ServiceResponse.model_validate(service)}


@router.get("/services", response_model=PaginatedResponse[ServiceResponse])
def list_services(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
    search: Optional[str] = Query(default=None),
    category: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
    delivery_model: Optional[str] = Query(default=None),
    pricing_model: Optional[str] = Query(default=None),
    is_featured: Optional[bool] = Query(default=None),
    is_active: Optional[bool] = Query(default=None),
    sort: str = Query(default="created_at"),
    order: str = Query(default="desc"),
    db: Session = Depends(get_db),
):
    items, total = ServiceCatalogService.list_services(
        db=db,
        page=page,
        page_size=page_size,
        search=search,
        category=category,
        status=status,
        delivery_model=delivery_model,
        pricing_model=pricing_model,
        is_featured=is_featured,
        is_active=is_active,
        sort=sort,
        order=order,
    )

    responses = [ServiceResponse.model_validate(s) for s in items]
    total_pages = math.ceil(total / page_size) if page_size > 0 else 0

    return {
        "data": responses,
        "pagination": PaginationMeta(
            page=page,
            page_size=page_size,
            total=total,
            total_pages=total_pages,
        ),
    }


@router.get("/services/slug/{slug}", response_model=DataResponse[ServiceResponse])
def get_service_by_slug(slug: str, db: Session = Depends(get_db)):
    service = ServiceCatalogService.get_service_by_slug(db, slug)
    return {"data": ServiceResponse.model_validate(service)}


@router.get("/services/{service_id}", response_model=DataResponse[ServiceResponse])
def get_service(service_id: uuid.UUID, db: Session = Depends(get_db)):
    service = ServiceCatalogService.get_service_by_id(db, service_id)
    return {"data": ServiceResponse.model_validate(service)}


@router.patch("/services/{service_id}", response_model=DataResponse[ServiceResponse])
def update_service(
    service_id: uuid.UUID,
    data: ServiceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ServiceCatalogService.update_service(db, service_id, data)
    return {"data": ServiceResponse.model_validate(service)}


@router.delete("/services/{service_id}", response_model=DataResponse[ServiceResponse])
def archive_service_by_delete(
    service_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ServiceCatalogService.archive_service(db, service_id)
    return {"data": ServiceResponse.model_validate(service)}


@router.post("/services/{service_id}/archive", response_model=DataResponse[ServiceResponse])
def archive_service(
    service_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ServiceCatalogService.archive_service(db, service_id)
    return {"data": ServiceResponse.model_validate(service)}


@router.post("/services/{service_id}/restore", response_model=DataResponse[ServiceResponse])
def restore_service(
    service_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ServiceCatalogService.restore_service(db, service_id)
    return {"data": ServiceResponse.model_validate(service)}


# Lead-Service Associations
@router.get("/leads/{lead_id}/services", response_model=DataResponse[List[LeadServiceResponse]])
def list_lead_services(
    lead_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    records = ServiceCatalogService.list_lead_services(db, lead_id)
    return {"data": [LeadServiceResponse.model_validate(r) for r in records]}


@router.post(
    "/leads/{lead_id}/services", response_model=DataResponse[LeadServiceResponse], status_code=201
)
def add_service_to_lead(
    lead_id: uuid.UUID,
    data: LeadServiceCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    record = ServiceCatalogService.add_service_to_lead(db, lead_id, data)
    return {"data": LeadServiceResponse.model_validate(record)}


@router.delete("/leads/{lead_id}/services/{service_id}")
def remove_service_from_lead(
    lead_id: uuid.UUID,
    service_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    success = ServiceCatalogService.remove_service_from_lead(db, lead_id, service_id)
    return {"data": {"success": success, "message": "Service association removed."}}
